src/engine/matching-engine.ts

```diff
@@ -28,93 +29,94 @@ const invalidateNonce = async (user: Hex, nonce: bigint) => {
   // TODO: Implement
 };
 
-let orderId = 0;
-
 export class MatchingEngine {
-  orders: LimitOrder[];
+  private buyOrders: Map<bigint, Map<string, LimitOrder>>;
+  private sellOrders: Map<bigint, Map<string, LimitOrder>>;
+  private orderIdToPrice: Map<string, bigint>;
 
   constructor(public readonly market: Market) {
-    this.orders = [];
+    this.buyOrders = new Map();
+    this.sellOrders = new Map();
+    this.orderIdToPrice = new Map();
   }
 
   async addOrder(order: LimitOrder) {
-    order.id = `${orderId++}`;
+    if (!order.id) {
+      order.id = randomBytes(16).toString('hex');
+    }
+
     if (!(await checkOrderSignature(order))) {
       throw new Error("Invalid order signature");
     }
-    offersOfUser[order.user] = offersOfUser[order.user] || {};
-    offersOfUser[order.user][order.id] = this.market;
-    this.orders.push(order);
+
+    const orderMap = order.order.limitOrderMaker ? this.buyOrders : this.sellOrders;
+    if (!orderMap.has(order.order.price)) {
+      orderMap.set(order.order.price, new Map());
+    }
+    orderMap.get(order.order.price)!.set(order.id, order);
+    this.orderIdToPrice.set(order.id, order.order.price);
+
     this.checkForPossibleSettles();
     return order.id;
   }
 
   getOrders() {
-    return this.orders;
+    return [...this.buyOrders.values(), ...this.sellOrders.values()];
```

Comment:
Can we expand + type this function to cover 

`Get{Buy,Sell}Orders{,atPrice}`?